### PR TITLE
Remove ip-matching dependency

### PR DESCRIPTION
**Pull request by Nitwel**
## What's Changed

- See title

## Tested Scenarios

- Added tests that should cover all edge cases

## Review Notes / Questions / Concerns

- Due to tests a code review should be sufficient

## Checklist

> Leave unchecked where not applicable

- [x] Tests added/updated
- [ ] Documentation PR created in [directus/docs](https://github.com/directus/docs) 
- [ ] OpenAPI updated
  - [ ] Local specs package (`@directus/specs`)
  - [ ] PR created in [directus/openapi](https://github.com/directus/openapi) 
- [ ] SDK (`@directus/sdk`) updated to reflect the changes
- [ ] Types (`@directus/types`) updated to reflect the changes
- [ ] GraphQL schema updated to reflect the changes
- [ ] System data (`@directus/system-data`) updated for changes to system collections/fields/relations
- [ ] Database migration added for schema/system changes
- [ ] Environment variables documented for new/changed config
- [ ] App translations added for new user-facing strings
- [ ] Security implications apply

### CODE REVIEW

File: api/src/services/policies.ts [hunk 1]
```diff
@@ -1,11 +1,61 @@
+import { isIP } from 'node:net';
 import { InvalidPayloadError } from '@directus/errors';
 import type { AbstractServiceOptions, MutationOptions, Policy, PrimaryKey } from '@directus/types';
 import { UserIntegrityCheckFlag } from '@directus/types';
-import { getMatch } from 'ip-matching';
+import { type IPv4, type IPv6, isValidCIDR, parse as parseIp } from 'ipaddr.js';
```
Comment (ComfortablyCoding) [inline api/src/services/policies.ts]: This breaks API startup as the package is not ESM.

```suggestion
import ipaddr from 'ipaddr.js';
```

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]
Since `ipaddr` is not an ESM package, the API currently fails to start. 

That said, is there any context on why we aren't leveraging `IpBlocklist`? It should already provide most of the functionality we need here and would prevent any drift between read/write.

**Comment by Nitwel** [comment]
I moved over the logic to IpBlocklist, should be a bit cleaner and reusable now.

File: packages/utils/node/ip-blocklist.ts
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.ts:1]: We should be able to rely on checks similar to the ones in [ipInNetworks](https://github.com/directus/directus/blob/2fb241fc01ad6c4f8953a5c286509752093a8465/packages/utils/node/ip-in-networks.ts#L9), coupled with the parse### functions when we need to determine validity. 

The newly added functions seem to largely duplicate their corresponding parse functions. There may still be some slight adjustments or additional validation needed for the is#### functions that aren't covered by `ipInNetworks`. We ideally stick to a uniform is/parse requirements where/when possible.

File: api/src/services/policies.ts [hunk 2]
```diff
@@ -1,11 +1,26 @@
 import { InvalidPayloadError } from '@directus/errors';
 import type { AbstractServiceOptions, MutationOptions, Policy, PrimaryKey } from '@directus/types';
 import { UserIntegrityCheckFlag } from '@directus/types';
-import { getMatch } from 'ip-matching';
+import { IpBlocklist } from '@directus/utils/node';
 import { clearSystemCache } from '../cache.js';
 import { clearCache as clearPermissionsCache } from '../permissions/cache.js';
 import { ItemsService } from './items.js';
 
+export function isIpAccessValid(value?: any[] | null): boolean {
```
Comment (ComfortablyCoding) [inline api/src/services/policies.ts:9]: Nit, `isValidIpAccess` would be inline with `assertValidIpAccess`

Comment (ComfortablyCoding) [reply api/src/services/policies.ts:9]: Not specific to changes here but keeping the old typing for array or null leads to improper validation.

No validation asserts that this is null or an array before passing it in, something like `1` leads to a `500`. We should handle possible cases accordingly.
```suggestion
export function isIpAccessValid(value?: unknown): boolean {
```

File: packages/utils/node/ip-blocklist.ts [hunk 3]
```diff
@@ -149,4 +144,45 @@ export class IpBlocklist extends BlockList {
...
+function isIpLessOrEqual(ipA: IP, ipB: IP) {
+	const bytesA = ipA.toByteArray();
+	const bytesB = ipB.toByteArray();
+
+	for (let i = 0; i < bytesA.length; i++) {
+		if (bytesA[i]! < bytesB[i]!) return true;
+		if (bytesA[i]! > bytesB[i]!) return false;
+	}
+
+	return true;
```
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.ts:187]: I believe we can simplify this via built in Buffer methods

```suggestion
	Buffer.compare(Buffer.from(start.toByteArray()), Buffer.from(end.toByteArray())) > 0;
```

File: api/src/services/policies.ts [hunk 4]
```diff
@@ -1,11 +1,26 @@
...
+	if (value === undefined) return false;
+	if (value === null) return true;
+	if (Array.isArray(value) && value.length === 0) return true;
+
+	for (let ip of value) {
+		if (typeof ip !== 'string' || ip.includes('*')) return false;
+		ip = ip.trim();
+
+		if (!IpBlocklist.isIP(ip) || !IpBlocklist.isRange(ip) || !IpBlocklist.isSubnet(ip)) return false;
```
Comment (ComfortablyCoding) [inline api/src/services/policies.ts:18]: This incorrectly returns false as it requires the ip to not be an address, a range and a subnet all at the same time.

Let's also add some test coverage here as well, since we have none.

File: packages/utils/node/ip-blocklist.ts [hunk 5]
```diff
@@ -149,4 +144,45 @@ export class IpBlocklist extends BlockList {
...
+		const ipMin = isIP(parts[0]!);
+		const ipMax = isIP(parts[1]!);
+		if (ipMin === 0 || ipMax === 0) return false;
+		if (ipMin !== ipMax) return false;
+		if (!isIpLessOrEqual(ipaddr.parse(parts[0]!), ipaddr.parse(parts[1]!))) return false;
+
+		return true;
+	}
+
+	static isIP(ip: string): boolean {
```
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.ts:161]: Nit: More aptly named `isAddress` and inline with others in the class

File: packages/utils/node/ip-blocklist.ts [hunk 6]
```diff
@@ -149,4 +144,45 @@ export class IpBlocklist extends BlockList {
 
 		return false;
 	}
+
+	static isRange(ip: string): boolean {
```
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.ts:148]: Nit: As it is cheap, lets add a jsdoc for these additional methods

File: api/src/services/policies.ts (hunk 4 above)
Comment (ComfortablyCoding) [reply api/src/services/policies.ts:18]: Nit: Worth using an `Blocklist.isNetwork(input)` (or similar named) function to determine this. That would enable it to be test backed and re-useable in the other uses where we should be checking it.

Comment (ComfortablyCoding) [reply api/src/services/policies.ts:18]: Nit: We can also add a `parseNetwork` equivalent which checks an input and calls relevant parse## function. Erroring if none of them match. That would reduce the custom code we have in `ipInNetworks` and `isDeniedIp`

File: packages/utils/node/ip-blocklist.ts [hunk 7]
```diff
@@ -101,15 +101,12 @@ export class IpBlocklist extends BlockList {
 	 * @throws {Error} Throws 'ERR_INVALID_SUBNET' if the input format is invalid
 	 */
 	parseSubnet(input: string): void {
```
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.ts:103]: `parseSubnet` unlike `getMatch` no longer accepts `10.0.0.0/ 24` and `10.0.0.0/+24` which while invalid was accepted and corrected by getMatch in ip-matching. Edge case, but if someone did accidentally add one of those it would be stored and after these changes start throwing errors.

File: .changeset/seven-geckos-poke.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@directus/api': patch
```
Comment (ComfortablyCoding) [inline .changeset/seven-geckos-poke.md:2]:
```suggestion
'@directus/api': patch
'@directus/utils': minor
```

File: packages/utils/node/ip-blocklist.test.ts [hunk 9]
```diff
@@ -269,4 +269,101 @@ describe('IpBlocklist', () => {
...
+	describe('IP masks', () => {
+		test.each(['10.0.0.0/255.255.255.0', '192.168.0.0/255.255.0.0', 'fe80::/ffff::'])('rejects mask %s', (ip) => {
+			expect(IpBlocklist.isIP(ip)).toBe(false);
+			expect(IpBlocklist.isRange(ip)).toBe(false);
+			expect(IpBlocklist.isSubnet(ip)).toBe(false);
+		});
+	});
+
+	describe('IP ranges', () => {
+		test.each(['10.0.0.1-10.0.0.5', '0.0.0.0-255.255.255.255', 'fe80::1-fe80::5', '::1-::2'])(
```
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.test.ts:338]: This can use some additional cases: range with same range address (e.g. `10.0.0.1-10.0.0.1`), range flip (e.g. `10.0.255.255-10.1.0.0`) etc

File: packages/utils/node/ip-blocklist.test.ts [hunk 10]
```diff
@@ -269,4 +269,101 @@ describe('IpBlocklist', () => {
...
+			'accepts range %s',
+			(ip) => {
+				expect(IpBlocklist.isIP(ip)).toBe(false);
+				expect(IpBlocklist.isRange(ip)).toBe(true);
+				expect(IpBlocklist.isSubnet(ip)).toBe(false);
+			},
+		);
+
+		test.each([
```
Comment (ComfortablyCoding) [inline packages/utils/node/ip-blocklist.test.ts:347]: Similar to the above could use more coverage: different ip types (e.g. `10.0.0.1-::1`) etc

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]
See comments below, I believe we can unify this a bit more. At the moment split over 3 different areas

File: api/src/services/policies.ts (hunk 2 above)
Comment (Nitwel) [reply api/src/services/policies.ts]: It's typed as `ip_access?: string[] | null;` so I'd say this type is valid?

File: packages/utils/node/ip-blocklist.ts (hunk 7 above)
Comment (Nitwel) [reply packages/utils/node/ip-blocklist.ts:103]: Yep, but I would assume that this would have failed later when comparing. We can add a breaking change warning though if you see this fit.

File: api/src/services/policies.ts (hunk 2 above)
Comment (ComfortablyCoding) [reply api/src/services/policies.ts]: That is from the db, the service/endpoint has no such restriction and will accept any type. Either way we should not be returning a 500 for an invalid payload/type.

File: packages/utils/node/ip-blocklist.ts (hunk 7 above)
Comment (ComfortablyCoding) [reply packages/utils/node/ip-blocklist.ts:103]: Yes, breaking change notice works instead of supporting it 🫡
